_packages/network/lib/auth/repository/auth_repository_impl.dart

```diff
@@ -0,0 +1,64 @@
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:google_sign_in/google_sign_in.dart';
+
+import 'auth_repository.dart';
+
+class AuthRepositoryImpl extends AuthRepository {
+  final FirebaseAuth auth;
+
+  AuthRepositoryImpl({
+    required this.auth,
+  });
+
+  Stream<User?> get authStateChange => auth.authStateChanges();
+
+  @override
+  Future<void> loginWithEmailAndPassword(
+      String username, String password) async {
+    try {
+      await auth.signInWithEmailAndPassword(
+          email: username, password: password);
+    } on Exception catch (e) {
+      throw Exception(e);
+    }
+  }
+
+  @override
+  Future<void> signInWithGoogle(String email) async {
+    // Trigger the authentication flow
+    final GoogleSignInAccount googleUser = await GoogleSignIn().signIn();
+
+    // Obtain the auth details from the request
+    final GoogleSignInAuthentication googleAuth =
+        await googleUser!.authentication;
+
+    // Create a new credential
+    final credential = GoogleAuthProvider.credential(
+      accessToken: googleAuth.accessToken,
+      idToken: googleAuth.idToken,
+    );
+
+    try {
+      await auth.signInWithCredential(credential);
+    } on Exception catch (e) {
+      throw Exception(e);
+    }
+  }
+
+  @override
+  Future<void> signOut() async {
+    await auth.signOut();
+  }
+
+  @override
+  Future<void> signUpWithEmailAndPassword(String email, String password) async {
+    try {
+      await auth.createUserWithEmailAndPassword(
+        email: email,
+        password: password,
+      );
+    } on Exception catch (e) {
+      throw Exception(e);
```

Comment:
What i meant was something like

```suggestion
}on FirebaseAuthException catch (e) {  
  if (e.code == 'network-request-failed') {
      throw Exception('No Internet Connection');
    } else if (code == "invalid-email" ||
        code == "wrong-password" ||
        code == "user-not-found") {
      throw Exception('Invalid username or Password');
    } else {
      throw Exception('Unknown Error');
    }
}
```

Reply:
I’ve created a gist with all codes incase you need.
https://gist.github.com/wizlif/16ae686441765ce6bd28bc1b4b6e941d